src/polhemus_tf_broadcaster.cpp

```diff
@@ -255,32 +224,31 @@ int main(int argc, char** argv) {
        The framecount is a more exact way of finding out the time:
        timestamp = framecount*1000/240 (rounded down to an int)* 
     */
-
+    
     for (i=0; i < 8; i++) {
-      if (broadcast_stations[i]) {
-	s = i + 1;
-	static tf2_ros::TransformBroadcaster br;
-	geometry_msgs::TransformStamped transformStamped;
-
-	// Header info
-	transformStamped.header.stamp = ros::Time::now();
-	transformStamped.header.frame_id = "polhemus_base";
-	transformStamped.child_frame_id = "polhemus_station_" + std::to_string(s);
+
+      static tf2_ros::TransformBroadcaster br;
+      geometry_msgs::TransformStamped transformStamped;
+
+      // Header info
+      transformStamped.header.stamp = ros::Time::now();
+      transformStamped.header.frame_id = "polhemus_base";
+      transformStamped.child_frame_id = "polhemus_station_" + std::to_string(i+1);
 	
-	// Set translation (inches)
-	transformStamped.transform.translation.x = stations[i].x;
-	transformStamped.transform.translation.y = stations[i].y;
-	transformStamped.transform.translation.z = stations[i].z;
-
-	// Set rotation
-	transformStamped.transform.rotation.x = stations[i].quaternion[3];
-	transformStamped.transform.rotation.y = stations[i].quaternion[0];
-	transformStamped.transform.rotation.z = stations[i].quaternion[1];
-	transformStamped.transform.rotation.w = stations[i].quaternion[2];
-
-	// Broadcast frame
-	br.sendTransform(transformStamped);
-      }
+      // Set translation (inches)
+      transformStamped.transform.translation.x = stations[i].x;
+      transformStamped.transform.translation.y = stations[i].y;
+      transformStamped.transform.translation.z = stations[i].z;
+
+      // Set rotation
+      transformStamped.transform.rotation.x = stations[i].quaternion[3];
```

Comment:
This looks suspiciously wrong. Is the convention in Polhemus really y,z,w,x? It should be either w,x,y,z (bot-frames convention for instance) or x,y,z,w (ROS convention). Please check the docs!

Reply:
@wxmerkt here i am extracting (w,x,y,z), i.e. (q[3], q[0], q[1], q[2]). 

The docs for the polhemus liberty that say the order should be (q[3], q[0], q[1], q[2]) (as in the code), see page 21 [here](http://polhemus.com/_assets/img/LIBERTY_User_Manual_URM03PH156-H.pdf). 

However, we have a problem using the following. 
```
// Set rotation
transformStamped.transform.rotation.x = stations[i].quaternion[3];
transformStamped.transform.rotation.y = stations[i].quaternion[0];
transformStamped.transform.rotation.z = stations[i].quaternion[1];
transformStamped.transform.rotation.w = stations[i].quaternion[2];
```
Rotation about the x-axis works as intended but about the other two axes the direction shown in RVIZ is in reverse. I have solved the problem (see recent commits) by changing the sign of q[0] and q[1], as follows.

```
// Set rotation
transformStamped.transform.rotation.x = stations[i].quaternion[3];
transformStamped.transform.rotation.y = -stations[i].quaternion[0];
transformStamped.transform.rotation.z = -stations[i].quaternion[1];
transformStamped.transform.rotation.w = stations[i].quaternion[2];
```
I currently have no idea why this happens. Could this be a problem in either `protocol.h` or `liberty.cpp`/`liberty.h`?

Reply:
Yeah that's what I meant - your order is wrong. Check what you are _assigning it to_. Your assigned order is xyzw (in the transformStamped) but you are assigning wxyz. Change the name of the variables you are assigning to and it should work.
And *never* change the sign of a quaternion (it's normalised).

Reply:
@cmower Seeing that the updated hacked solution again involves flipping signs in components of the quaternion, I recommend having a look at how quaternions work [and why flipping signs is a bad idea, e.g. RSS lecture slides, wikipedia]. The issue is not in the order in which you are getting it from the liberty quaternion array (3,0,1,2 - WXYZ) but where you are assigning it to on the left hand-side (xyzw). Just changing the ".x/.y/.z/.w" on the left hand side (transformStamped) to the correct order will solve your problem. Happy to stop by 1.30 after lunch for a face to face conversation